src/openfermion/utils/_sparse_tools.py

```diff
@@ -703,58 +703,63 @@ def sparse_eigenspectrum(sparse_operator):
-            or, a scipy.sparse.csc matrix representing a density matrix.
+        operator(scipy.sparse.spmatrix or scipy.sparse.linalg.LinearOperator):
+            The operator whose expectation value is desired.
+        state(numpy.ndarray or scipy.sparse.spmatrix): A numpy array
+            representing a pure state or a sparse matrix representing a density
+            matrix. If `operator` is a LinearOperator, then this must be a
+            numpy array.
 
     Returns:
-        A real float giving expectation value.
+        A complex number giving the expectation value.
 
     Raises:
         ValueError: Input state has invalid format.
     """
-    # Handle density matrix.
-    if state.shape == sparse_operator.shape:
-        product = state * sparse_operator
+
+    if isinstance(state, scipy.sparse.spmatrix):
+        # Handle density matrix.
+        if isinstance(operator, scipy.sparse.linalg.LinearOperator):
+            raise ValueError('Taking the expectation of a LinearOperator with '
+                             'a density matrix is not supported.')
+        product = state * operator
         expectation = numpy.sum(product.diagonal())
 
-    elif (state.shape == (sparse_operator.shape[0], 1) or
-          state.shape == (sparse_operator.shape[0], )):
+    elif isinstance(state, numpy.ndarray):
         # Handle state vector.
-        if scipy.sparse.issparse(state):
-            expectation = (state.getH() * sparse_operator * state)
-        else:
-            expectation = numpy.dot(numpy.conj(state.T),
-                                    sparse_operator.dot(state))
-        if expectation.shape != ():
-            expectation = expectation[0, 0]
+        expectation = numpy.dot(numpy.conjugate(state), operator * state)
+
     else:
         # Handle exception.
-        raise ValueError('Input state has invalid format.')
+        raise ValueError(
+                'Input state must be a numpy array or a sparse matrix.')
 
     # Return.
     return expectation
 
 
-def variance(sparse_operator, state):
+def variance(operator, state):
     """Compute variance of operator with a state.
 
     Args:
-        state: scipy.sparse.csc vector representing a pure state,
-            or, a scipy.sparse.csc matrix representing a density matrix.
+        operator(scipy.sparse.spmatrix or scipy.sparse.linalg.LinearOperator):
+            The operator whose expectation value is desired.
+        state(numpy.ndarray or scipy.sparse.spmatrix): A numpy array
+            representing a pure state or a sparse matrix representing a density
+            matrix.
 
     Returns:
-        A real float giving the variance.
+        A complex number giving the variance.
 
     Raises:
         ValueError: Input state has invalid format.
     """
-    return (expectation(sparse_operator ** 2, state) -
-            expectation(sparse_operator, state) ** 2)
+    return (expectation(operator ** 2, state) -
```

Comment:
A thought as I'm reading this is that powers of linear operators should ideally not square the terms in the operator, but rather result in a new linear operator that applies the original twice to a state.  We may not need this optimization now, but if someone starts writing a lot of routines calling higher powers of operators, we will.

Reply:
Squaring a LinearOperator gives an operator of type
```
<class 'scipy.sparse.linalg.interface._PowerLinearOperator'>
```
I think scipy already performs this optimization, but not sure.

Reply:
I verified this by inspecting the source code: https://github.com/scipy/scipy/blob/master/scipy/sparse/linalg/interface.py.

LinearOperators can only be raised to non-negative integer powers.

Reply:
Awesome, didn't realize LinearOperator did that by default